beluga/include/beluga/views/sample.hpp

```diff
@@ -0,0 +1,210 @@
+  /**
+   * It uses std::discrete_distribution to sample from the range.
+   */
+  template <
+      class Range,
+      class Weights,
+      class URNG = typename ranges::detail::default_random_engine,
+      std::enable_if_t<ranges::range<Range> && ranges::range<Weights>, int> = 0>
+  constexpr auto operator()(Range&& range, Weights&& weights, URNG& engine = ranges::detail::get_random_engine())
+      const {
+    static_assert(ranges::input_range<Weights>);
+    using result_type = ranges::range_difference_t<Range>;
+    auto w = ranges::views::common(weights);
+    auto distribution = std::discrete_distribution<result_type>{ranges::begin(w), ranges::end(w)};
+    return sample_view{ranges::views::all(std::forward<Range>(range)), std::move(distribution), engine};
+  }
+
+  /// Overload that implements the sample algorithm for non-weighted ranges.
+  /**
+   * It uses std::uniform_int_distribution to sample from the range.
+   */
+  template <
+      class Range,
+      class URNG = typename ranges::detail::default_random_engine,
+      std::enable_if_t<ranges::range<Range>, int> = 0,
+      std::enable_if_t<!is_particle_range_v<Range>, int> = 0,
+      std::enable_if_t<!ranges::range<URNG>, int> = 0>
+  constexpr auto operator()(Range&& range, URNG& engine = ranges::detail::get_random_engine()) const {
+    using result_type = ranges::range_difference_t<Range>;
+    auto distribution =
+        std::uniform_int_distribution<result_type>{0, static_cast<result_type>(ranges::size(range) - 1)};
+    return sample_view{ranges::views::all(std::forward<Range>(range)), std::move(distribution), engine};
+  }
+
+  /// Overload that handles particle ranges.
+  template <
+      class Range,
+      class URNG = typename ranges::detail::default_random_engine,
+      std::enable_if_t<ranges::range<Range>, int> = 0,
+      std::enable_if_t<is_particle_range_v<Range>, int> = 0,
+      std::enable_if_t<!ranges::range<URNG>, int> = 0>
+  constexpr auto operator()(Range&& range, URNG& engine = ranges::detail::get_random_engine()) const {
+    return (*this)(ranges::views::all(range), beluga::views::weights(range), engine);
+  }
+
+  /// Overload that unwraps the engine reference from a view closure.
+  template <class Range, class URNG, typename std::enable_if_t<ranges::range<Range>, int> = 0>
+  constexpr auto operator()(Range&& range, std::reference_wrapper<URNG> engine) const {
+    return (*this)(ranges::views::all(std::forward<Range>(range)), engine.get());
+  }
+
+  /// Overload that returns a view closure to compose with other views.
+  template <class URNG, std::enable_if_t<!ranges::range<URNG>, int> = 0>
+  constexpr auto operator()(URNG& engine) const {
+    return ranges::make_view_closure(ranges::bind_back(sample_fn{}, std::ref(engine)));
+  }
+};
+
+}  // namespace detail
+
+/// [Range adaptor object](https://en.cppreference.com/w/cpp/named_req/RangeAdaptorObject) that
+/// will ramdomly sample with replacement from an input range.
```

Comment:
@nahueespinosa typo: 
```suggestion
/// will randomly sample with replacement from an input range.
```